### ISSUE TITLE
$each with Vuelidate next is no longer a property

### ISSUE THREAD

**Issue by robyedlin** [vuelidate/vuelidate#781]
Is there an example for using $each with Vuelidate next? I don't see the $each property in the validations object anymore.

**Comment by dobromir-hristov** [comment]
That is no longer available. Use component validations for your objects. The parent will collect the validation results.

Check the docs

**Comment by robyedlin** [comment]
Thanks. FYI, there is a reference to the $each keyword here, which is why I asked: https://vuelidate-next.netlify.app/examples.html#collections-validation

**Comment by robyedlin** [comment]
Also, the terminology is a little unclear, I think that's why I didn't notice it in the docs.

**Comment by shentao** [comment]
Oh crap! I think this is some leftover from when we started migrating the docs. In the advanced usage, there is a mention that `$each` has been removed. We must have missed it in that part. We haven’t yet focused on the docs. :(

**Comment by robyedlin** [comment]
No problem, thanks!

**Comment by klausXR** [comment]
> That is no longer available. Use component validations for your objects. The parent will collect the validation results.
> 
> Check the docs

I have a question about the removal of this property.

My application supports multiple languages and allows users to update them. For example, when they want to create a product, they should provide its name in multiple languages.

They have similar structure to this

```js
const product = {
    locales: [
        { locale: "en", title: "Product title in English", description: "Product description in English" },
        { locale: "es", title: "Product title in Spanish", description: "Product description in Spanish" },
    ]
}
```

For the validations object, I'm using this schema

```
validations: {
    product: {
        locales: {
            $each: {
                title: { required },
                description: { required }
            }
        }
    }
}
```

I'm doing something else in my templates to model this, but this was the simplest example I could come up with

```html
<!-- Button to select locale to edit -->
<button v-for="(locale, index) in product.locales" :key="locale" @click="setSelectedLocale(index)">{{locale}}</button>

<input type="text" v-model="product.locales[selectedLocale].title">

<textarea v-model="product.locales[selectedLocale].description"></textarea>

<script>
    export default {
        data() {
            selectedLocale: 0
        },
        methods: {
          setSelectedLocale(index) {
            this.selectedLocale = index
          }
        }
    }
</script>
```

As can be seen from this example, I don't actually need a nested component for validations ( I would need a component per field ) and I'm confused as to how I should approach this situation.

I think there are many situations where one would like to validate all objects in an array and not want to create additional components.

**Comment by robyedlin** [comment]
I'm not sure about this choice either, but in some ways using components can be cleaner.

The way I've compromised is to use a reusable renderless component that accepts props for fields and rules, and sends a slot prop back with the validations. That will give you the validations and you don't have to create a separate component every time you want nested validations.

For example:

```
v-nested-form(v-for="(item, index) in items" :fields="item.fields" :rules="item.rules" v-slot="{ itemV$ }")
    // now you can can use v-model using index and itemV$ to validate
    button
    input
    textarea
```

**Comment by klausXR** [comment]
> I'm not sure about this choice either, but in some ways using components can be cleaner.
> 
> The way I've compromised is to use a reusable renderless component that accepts props for fields and rules, and sends a slot prop back with the validations. That will give you the validations and you don't have to create a separate component every time you want nested validations. The parent vuelidate instance will pick up errors from nested instances too (scoped nested validations are coming).
> 
> For example:
> 
> ```
> v-nested-form(v-for="(item, index) in items" :fields="item.fields" :rules="item.rules" v-slot="{ itemV$ }")
>     // now you can can use v-model using item, index, and itemV$ to validate
>     button
>     input
>     textarea
> ```

Yes, I do agree that it is useful in certain scenarios and it should be touted as a best practice, to be used wherever fit. However, you should not be forced to write code this way. This is a model-based object validator, it shouldn't matter how your templates look. So, how do you feel about having to use the renderless component just so that you comply with the requirements of the library? 

I honestly do not feel very happy about it. I do not feel like Vuelidate belongs to one of the layers that dictates how I should structure my code, in the same way that, for example, Nuxt is. 

I do not build applications on top of Vuelidate, I built applications with it. So, I should be able to use it with more freedom than what I have seen to be described so far.

**Comment by robyedlin** [comment]
@shentao Is there any documentation on this design choice?

**Comment by ajdvoynos** [comment]
Please update the docs 😣 I just spent half a day trying to figure out why $each wasn't working. I tried to find the page on github to send a pull request, but I don't think the vnext docs are in github?

I also agree that having to create extra components just because of vuelidate is burdensome, and that's exactly what I'm in the middle of right now.

**Comment by klausXR** [comment]
Well, on the readme under `Features & characteristics:` it says "Decoupled from templates". I guess no  more

**Comment by dobromir-hristov** [comment]
I am going to close this for now. `$each` is probably not coming back in the form it was. In our experiments, it proved that it is much more flexible to use components to validate collections of data. 

We will be exposing other methods to validate collections for simple cases, as well as improved docs, but for now try to use nested forms and the new `$scope` feature - https://vuelidate-next.netlify.app/advanced_usage.html#validation-scopes

It is sort of true, that this expects some form of template coupling, but it expands to just having a child component form.

**Comment by shentao** [comment]
Sorry to hear that this change is causing that much trouble and confusion. When discussing it internally, we figured that using nested components will play well with general Vue best practices. Where, when you have a list of input sets, you usually want to extract it into a new component anyway. 

That’s why we focused on the aspect where nested components will push their errors to the closest vuelidate-enabled ancestor component. It also solved one of the most frustrating problems – usage within dynamic forms, where the root component doesn’t know the exact data structure (nor it should), and the child components need to have their own validations. Collecting those validations and then sending them back up has been a pain – I’m speaking from my personal work experience as well as what I have seen when I was consulting.
Having to specify the validations that are several layers deep and then create templates with nested `v-for` was hard. Trying to sync child validations with parents (especially with v-for going on) was even harder and required a lot of coupling between those components, which is again a bad practice. It also often required a lot of additional code to handle merging those validation results.

>  I do not feel like Vuelidate belongs to one of the layers that dictates how I should structure my code, in the same way that, for example, Nuxt is.

And this is exactly what we wanted to fix with this change!

With that in mind, we figured that introducing automatic validation collecting would solve 2 problems at once. From my experience it did. In a large project (2k+ components) with forms being generated from a schema, this allowed us to remove hundreds of lines of additional code related to handling nested Vuelidate validations.

Seriously, please give it a try! 

We are also exploring a way to have validators return more than just a boolean. This should come in handy with async validators, where we could have the API response affect the error message. @dobromir-hristov is already working on that as we speak.
Additionally, it might bring back some of the early ideas for how we can replace `$each`. Namely `every` and `some` helpers. Which technically you can already write if you want, though their functionality will be somewhat limited before the above functionality is introduced.

This is how they were supposed to work.
```js
// where userEmails is a list of strings [ 'email1', 'email2' ]
const rules = {
  userEmails: every(email)
}

// or in case of user objects
const rules = {
  users: every({ 
    email: { email, required }
  })
}
```

Technically, you can already implement a helper like that. In the first scenario it would look like this:
```js
const every = validator => list => list.every(element => validator.$validator(element))
```

The second scenario would require some additional parsing of the passed schema declaration, but I’m sure you can figure that out given the basic scenario above.

The only drawback now is that you would only get a `true/false` from this, which would not inform which exact element in the list is invalid. But this is something that can be fixed with the functionality Dobromir is working on. 

Thus, we can’t promise you that `$each` will be back in the form it used to be, but you can expect that there will be some helpers to help you apply validators on arrays.
However, we still stand that in the case of more advanced forms, one should use a nested component. If not for Vuelidate’s sake, definitely do it for future maintenance sake.

Re-opening to hear back your input on the direction we are taking here.

**Comment by alperencaliskan** [comment]
Thanks for code samples @shentao. I understand that at worst we can put arrays in child components to validate. But how can one use it in custom slot in components? For example i use buefy table component for product list. Each row is validated.

```vue
<b-table
    :data="Array.from(Object.values(p.products.$each.$iter))"
    detailed
    detail-key="description"
    ref="table"
    :show-detail-icon="false"
  >
    <b-table-column
      field="description"
      width="auto"
      label="Product Name"
      v-slot="props"
    >
      <b-field
        type="is-info"
      >
        <input
          class="input"
          v-model="props.row.description.$model"
          :class="{
            'is-danger': props.row.description.$error
          }"
          maxlength="60"
        >
      </b-field>
    </b-table-column>
    .
    .
    .
<b-table>
```

**Comment by shentao** [comment]
@alperencaliskan Hmm... If you can replace that `<input>` with a component and say pass `value` there, then you would only need to do something like this:
```js
setup(props) {
  const { value, rules } = toRefs(props)
  useVuelidate({ value }, { value: rules.value })
}
```

And then in the parent component where you have the table, you only need:
```js
setup () {
  return { v$: useVuelidate() }
}
```

To get all the error messages from child components you can then do `v$.$errors` in the parent component. 
The risk here is that the error messages might look the same, so you might want to enhance the built-in error messages with a piece of extra information on which field is actually invalid. 

Unless of course, you want to handle that error message on a per-row basis. Then you can update the first snippet to be like this:
```js
setup(props) {
  const { value, rules } = toRefs(props)
  return {
    v$: useVuelidate({ value }, { value: rules.value })
  }
}
```

I hope that helps!

**Comment by alperencaliskan** [comment]
@shentao Thank you for your response. I have created a generic component for slot validations. And can collect validation errors from children. Seems to work for now.

```vue
<template>
  <div>
    <slot :vscope="v$"></slot>
  </div>
</template>

<script lang="ts">
import { defineComponent, watch, computed,reactive } from "@vue/composition-api";
import { useVuelidate } from "@vuelidate/core";


export default defineComponent({
  props: {
    value: { required: true, type: [Object, Array, Number, String] },
    validation: { required: true, type: Object },
  },
  components: {
  },
  setup: (props, ctx) => {
    const model = reactive(props.value);

    const rules = computed(() => ({
        ...props.validation,
    }));

    watch(model, (oldVal, newVal) => {
      ctx.emit("input", model.value);
    });

    return {
      model,
      v$: useVuelidate(rules, model, { $scope: true}),
    }
  },
});
</script>
```
Using in buefy table column slot
```vue
<b-table-column
      field="description"
      width="auto"
      label="Product Name"
      v-slot="props"
    >
      <b-field
        type="is-info"
      >
        <vuelidate-component
          v-model="props.row"
          v-slot="v"
          :validation="rules" <!-- defined in setup -->
        >
          <input
            class="input"
            v-model="v.vscope.name.$model"
            :class="{
              'is-danger': v.vscope.name.$error
            }"
            maxlength="60"
          >
        </vuelidate-component>
      </b-field>
    </b-table-column>
```

**Comment by shentao** [comment]
Glad I could help!

**Comment by alperencaliskan** [comment]
We need array validation back. :sob:

**Comment by shentao** [comment]
We are exploring new solutions for this problem. :)

**Comment by rcovell** [comment]
We are building a library of reusable components/atoms/molecules that will be consumed via a plugin. One component `ContactEditForm` contains an "array" of `PhoneNumber` (also components). This scenario lends itself to the suggestion above, to let components validate themselves (all for it).

Long story, short. Before we new that array validation was removed we tried to perform the actual validation from the App that consumed components from the plugin. The App sets everything up and passes `v$` down to child components to listen for validation errors (largely they are dumb, we wanted them to be). Once we saw that arrays were a no go, we moved validation into the plugin. Didn't think it would be that hard as they were already components (just pass them their rules and go).

Once the app consumed any of the components that do validation we get the following warning:

`[Vue warn]: Error in data(): "TypeError: Cannot read property 'uid' of null"`

This appears when any plugin component, `ContactEditForm`, calls: `useVuelidate`. Now this might be a webpack problem (suggestions), or it might be that we shouldn't have our plugin coupled to vuelidate at all. All in all we are still wrestling with how to validate arrays (of nested components, phone numbers) that are delivered from a plugin.

If anyone has any suggestions on either of the above issues please let me know.

**Comment by dobromir-hristov** [comment]
`[Vue warn]: Error in data(): "TypeError: Cannot read property 'uid' of null"` - this usually means the Vue instance was null, when `useVuelidate` was called. Do you by any chance call it outside of a `setup` function or after some `await`? Maybe you should move this to a new issue though.

**Comment by rcovell** [comment]
Moved my question to: [https://github.com/vuelidate/vuelidate/issues/820](url)

Original thought process was solving my issue of validating N PhoneNumbers. Above warning was a side affect of trying to do that.

**Comment by o3-steven** [comment]
> I do not feel like Vuelidate belongs to one of the layers that dictates how I should structure my code, in the same way that, for example, Nuxt is.

> And this is exactly what we wanted to fix with this change!

> @alperencaliskan Hmm... If you can replace that `<input>` with a component and say pass `value` there, then you would only need to do something like this:

Vuelidate just dictated that I create a component so that I can get validation for an array of items that would at most add 4 fields to the main form. I see the reasoning behind why creating components is good for an array of items especially from a Vue standpoint. But it's creating a lot of code for these smaller cases. 

Creating components is considered best practice in my mind when a region of code starts to pass a certain complexity and I think sort of what you explain here:

> However, we still stand that in the case of more advanced forms, one should use a nested component. If not for Vuelidate’s sake, definitely do it for future maintenance sake.

Would be nice to find some sort of middle ground for the cases that aren't so advanced and may not need to be put into a component yet. 

With all of that being said so far have loved the new version. Thanks for all the work y'all have put into this

**Comment by anorudes** [comment]
Thanks for the new version.
We are rewriting a large project with a huge number of forms in vue 3, and unfortunately the changes in vuelidation are the hardest part. Yes, we try to divide into different components, but often it makes no sense. Sometimes it is an array of phones, an array of postal addresses, and so on. Creating a separate component for this seems like an overcomplication.
I really hope that `$each` will come back. Many other libraries, for example, `yup`, have the same functionality.
It is now too difficult to take and rewrite all forms.
Maybe return `$each` and make it deprecated? To rewrite all forms gradually.

**Comment by dobromir-hristov** [comment]
I understand the frustration, but re-creating it back may not be as easy as it looks. There are a few outstanding issues that we need to fix before we can tackle $each.

If someone wants to give it a try, be my guest.
Development is slow, as I am the only one atm working on the library in late evenings after work hours, which is often not ideal.

I suppose we would have to basically detect the presence of $each in validators and try to iterate over the related state. That may require changes to the computed properties that detect error state and stuff.

**Comment by shentao** [comment]
I wonder, would it be possible to combine the `every` implementation idea mentioned earlier in this [thread](https://github.com/vuelidate/vuelidate/issues/781#issuecomment-774663432), with the ability for validators to return non-boolean results? 

@anorudes what level of granularity in the results would you require? 

```js
// simple version where we have a validator for the whole object
const every = validator => list => list.every(element => validator.$validator(element))

// validator that combines two validators
const phoneAndEmailValidator = ({ phone, email }) => emailValidator(email) && phoneValidator(phone)

// usage
{ 
  users: {
    allHaveEmailsAndPhones: every(phoneAndEmailValidator)
  }
}
```

The above would create a generic validation error for the whole list if at least one of the elements fails the check.
This is also something you could probably use right now.

In the future, assuming we could use the validators returning non-boolean values and switched `<Array>.every` to `<Array>.filter` to find which elements fail and then potentially pass them into `$message` we could get more detailed feedback on to which elements actually failed.

Maybe we could build in helper that allows mapping validators into obj properties allowing for usage presented below. Such a helper could potentially be constructed from the functions available in Vuelidate itself.
```js
  users: {
    allHaveEmailsAndPhones: every(validateObject({
      phone: { phoneValidator, required },
      email: { emailValidator, required }
    }))
  }
```

**Comment by dobromir-hristov** [comment]
We already support rich response objects for validators, so I suppose such a helper could return which properties are invalid. But that would rerun the validation on the entire array of items in each change, I think...

**Comment by shentao** [comment]
Yes it would. But that's not that is likely not a big problem when used for lists of elements that are not worth a dedicated component.

**Comment by ajdvoynos** [comment]
You normally also want to know which element in the list is the one that's failing validation, so you can put a red border or a message, how would you achieve that with the `every` solution above?

**Comment by anorudes** [comment]
@shentao This is an interesting idea, but we also show the error label text next to the input and highlight it in red. How to solve this with `every`?

**Comment by dobromir-hristov** [comment]
We can return an object with each invalid property, that you can access with `v.name.validator.$response`, along the necessary $valid key.

I can prototype something simple later today.

**Comment by robyedlin** [comment]
I may have a compromise. People seem not to like to have to create new components to replace `$each`, so what about creating a headless component in vuelidate that returns a slot prop?

Sample syntax:

```
ValidateEach( // headless component provided by Vuelidate
    v-for="(arrayField, index) in arrayFields"
    :field="arrayField"
    :rules="arrayRules"
    #default="{ vEach$ }")
         TextInput
             v-model="arrayField"
             :v$="vEach$")
```

**Comment by dobromir-hristov** [comment]
That might work... I was thinking from a data driven perspective, did not think about a component. Will play with it as well.

**Comment by anorudes** [comment]
Any updates? This problem prevents us from switching to vue 3. Our application has a huge number of forms (ERP system with dozens of forms).

**Comment by dobromir-hristov** [comment]
@anorudes I played with this a bit. Let me know how something like this feels.

### Create a EachProvider.vue component. We may ship it with Vuelidate

```vue
<script>
import { useVuelidate } from '@vuelidate/core'

export default {
  name: 'EachProvider',
  props: ['rules', 'state', 'options'],
  setup (props, { slots }) {
    const v = useVuelidate(props.rules, props.state, props.options)
    return () => slots.default({ v: v.value })
  }
}
</script>
```

### Define a collection and rules in your parent component
```js
// Parent.vue
setup() {

    // ... other state, rules, validations etc.
    const v = useVuelidate(rules, state)
    const collection = reactive([])
    const collectionRules = {
      name: { required }
    }

    function addToCollection () {
      collection.push({ name: '' })
    }

  return {
    collection,
    collectionRules,
    addToCollection
  }
}
```

### Use the component in your template

```vue
<EachProvider
      v-for="(item, index) in collection"
      :key="index"
      :state="item"
      :rules="collectionRules"
>
      <template #default="{ v }">
        <input
          v-model="v.name.$model"
          type="text"
        >
      </template>
</EachProvider>
```

**Comment by dobromir-hristov** [comment]
I know its not like just defining `$each` in your validations, but you dont need to create a dedicated component for each collection.

Using `every` would re-run the validator for every property in the entire collection, which I am not sure is performant. 

I also tried implementing it so it can be used in the below syntax, and it was not only very complex, but Vuelidate would throw a fuss because it could not parse it initially. 

```js
const rules = {
  people: every({
    name: { required },
    email: { email }
  })
}

const state = {
  people: [{
    name: 'John',
    email: 'john@mail.com'
  }]
}
```

**Comment by robyedlin** [comment]
@dobromir-hristov Yes, I think this is the right direction: https://github.com/vuelidate/vuelidate/issues/781#issuecomment-857050488

**Comment by IlCallo** [comment]
I'd like to highlight this use case I found myself in: https://github.com/vuelidate/vuelidate/issues/676#issuecomment-729844772

Using a wrapper component won't do the trick when you aren't in control of the template of component you're using (eg. many component frameworks out there), unless you opt-out of existing preset component and write the implementation yourself. And even if you do, you'll need to create a new component to do it.
This mostly happens when dealing with slots and dynamic slots

Since when I wrote that comment, I started to like the new way of doing this (encapsulating the validation into a child component), but for easier scenarios it still seems an overkill

**Comment by dobromir-hristov** [comment]
> I'd like to highlight this use case I found myself in: https://github.com/vuelidate/vuelidate/issues/676#issuecomment-729844772
> 
> Using a wrapper component won't do the trick when you aren't in control of the template of component you're using (eg. many component frameworks out there), unless you opt-out of existing preset component and write the implementation yourself. And even if you do, you'll need to create a new component to do it.
> This mostly happens when dealing with slots and dynamic slots
> 
> Since when I wrote that comment, I started to like the new way of doing this (encapsulating the validation into a child component), but for easier scenarios it still seems an overkill

Would the small wrapper component I posted above do the trick for you?

**Comment by robyedlin** [comment]
@IlCallo Yeah, re: https://github.com/vuelidate/vuelidate/issues/781#issuecomment-857432397, the new strategy looks unusable if the component expects an array and does its own v-for logic internally (although I don't exactly see where you are using validations in the example you gave).

**Comment by IlCallo** [comment]
> Would the small wrapper component I posted above do the trick for you?

Well, this is essentially the same workaround I ended up using, but I used a custom component instead of a generic wrapper.
I still had to increase my code complexity and opt-out of some features to apply this.
Thinking a bit more about it, this could actually work, but I need to test it.

> although I don't exactly see where you are using validations in the example you gave

In Vuelidate 1 I used `$each` to validate objects, in Vuelidate 2 I had to ditch the usage of cells dynamic slot and provide a component for the whole row which manages Vuelidate rules for that object.

Here's pseudo code on how I made it work now
<summary>
```ts
    <q-table
      :data="books"
      :columns="columns"
    >
      <!--
          Here I only need to customize a specific cell column (slot name would be something like `#body-cell-isbn`), but I cannot provide it as a normal slot contents as I won't be able to apply validation to it.
          BookTr is the custom component holding the object validation rules
        -->
      <template #body="props">
        <book-tr
          :key="props.row.id"
          :book="props.row"
          :props="props"
        />
      </template>
    </q-table>
```
</summary>

This could be a way to make it work using the wrapper component, but I need to check if it will actually works
<summary>
```ts
    <q-table
      :data="books"
      :columns="columns"
    >
      <template #body-cell-actions="props">
        <EachProvider
          :key="props.row.id"
          :state="props.row"
          :rules="bookCollectionRules"
        >
          <template #default="{ v }">
            <input
              v-model="v.isbn.$model"
              type="text"
            >
          </template>
        </EachProvider>
      </template>
    <q-table>
```
</summary>

Anyway, this seems like an advacement torwards the solution, so thank you for taking the time to work on this issue :)

**Comment by dobromir-hristov** [comment]
OK good news. I managed to create a very simple to use helper, that sort of gets `$each` back and is totally optional. The biggest CON, is that it will re-run all validators, for all items in the collection. That shouldnt be an issue for a small and simple case, but be warned. 

If you need to validate a big array of objects, just use a child component or the `EachProvider.vue` (srsly need to pick a better name, if I release it).

This is by no means complete, I still need to add tests and try with Vue 2 or the Options API.

## Link to live preview

https://codesandbox.io/s/vuelidate-2-with-every-helper-lsz1b

## Proposed Usage

```js
// inside setup
    const array = ref([])

    let v$ = useVuelidate({
      array: {
        $each: every({
          name: {
            required,
            minLength: minLength(10)
          }
        })
      }
    },
    { array })
    return { v$, array }
```

## Using in template
```vue
<div
      v-for="(input, index) in array"
      :key="index"
    >
      <input
        v-model="input.name"
        type="text"
      >
      <div v-for="message in v$.array.$each.$message[index]">
        <div v-for="string in message">
          {{ string }}
        </div>
      </div>
    </div>
```

This is what `v$.array.$each.$message` will hold:

```js
[
   {
      "name":{
         "required":"Value is required",
         "minLength":"This field should be at least 10 long"
      }
   },
   {
      "name":{
         "required":"Value is required",
         "minLength":"This field should be at least 10 long"
      }
   }
]
```

**Comment by anorudes** [comment]
Thank you, @dobromir-hristov. This will help us migrate our large application to vue 3. After the migration, we will refactor all the forms gradually and get rid of $each.

**Comment by robyedlin** [comment]
Cool! I think there is a good plan here!

@IlCallo you will be able to use $each but with hundreds of rows you may see degraded performance.

**Comment by robyedlin** [comment]
@dobromir-hristov what about calling the component ValidateEach

**Comment by dobromir-hristov** [comment]
The PR is up - https://github.com/vuelidate/vuelidate/pull/880

Anyone interested, give it a look :) It introduces a new `@vuelidate/components` package that holds the `ValidateEach` component, along with a new `forEach` helper, that is a more data-driven alternative. Both have PROs and CONs, explained in the new docs chapter.

**Comment by robyedlin** [comment]
I think this is a good compromise!

**Comment by IlCallo** [comment]
Cannot test it out right now, but the PR looks good to me and now there's surely enough flexibility to cover the cases we have been discussing

**Comment by iamunadike** [comment]
Hello sir, I am having issue with this I have giving $response.$errors is undefined 
When I call it in the template, the $response always shows 
```
setup() {
        const rules = {
            ticket: {
                serial: { required },
            },
            tasks: {
                $each: helpers.forEach({
                    tracker: { required },
                    description: { required },
                    service_ids: { required }
                })
            }
        };
        const state = reactive({
            ticket: {
                serial: "",
                status: "",
                tasks: []
            },
            tasks: [{
                tracker: "",
                description: "",
                rating: 0,
                service_ids: [],
                //workers: []
            }],
            payment: {
                type: "",
                deposit: 0.00,
                total: 0.00
            },
        })
        const v$ = useValidate(rules, state)

        return { v$, state }
```

**Comment by dobromir-hristov** [comment]
> Hello sir, I am having issue with this I have giving $response.$errors is undefined 
> When I call it in the template, for the tthis.v$.tasks.$each.$response.$errors.
> 
> import { helpers, required } from "@vuelidate/validators";
> 
> ```
> setup() {
>         const rules = { 
>             ticket: {
>                 serial: { required },
>             },
>             tasks: {
>                 $each: helpers.forEach({
>                     tracker: { required },
>                     description: { required },
>                     service_ids: { required }
>                 })
>             }
>         };
>         const state = reactive({
>             ticket: {
>                 serial: "",
>                 status: "",
>                 tasks: []
>             },
>             tasks: [{
>                 tracker: "",
>                 description: "",
>                 rating: 0,
>                 service_ids: [],
>                 //workers: []
>             }],
>             payment: {
>                 type: "",
>                 deposit: 0.00,
>                 total: 0.00
>             },
>         })
>         const v$ = useValidate(rules, state)
> 
>         return { v$, state }
> ```
> 
>  Tracker {
>   "$params": {},
>   "$pending": false,
>   "$invalid": true,
>   "$response": "TypeError: can't convert undefined to object"
> }

Please open a new ticket with the issue and a reproduction in codesandbox. I will look into it asap.

**Comment by rashad-yusifov** [comment]
Array validations in Vuelidate 3, i couldn't find any solution about this, but i've wrote some code it's working in dynamic Array validations.  let's try.
 

**VUE CODE**




```

import useVuelidate from "@vuelidate/core";
import {
  required,
  requiredIf,
  maxLength,
  minLength,
} from "@vuelidate/validators";


export default {

 data(){
   return {
       addresses: [
          {address_type: "", full_address: ""}
      ] 
   }

 }
validations(){
let self = this;
return {
addresses: [
          ...(() => {
             let validatonsArray = [
              {
                address_type: {
                  required,
                },
                full_address: {
                  required,
                },
              },
            ];
            self.addresses.forEach((item, index) => {
               if(index > 0){
                    validatonsArray.push({
                                    address_type: {
                                      required,
                                    },
                                    full_address: {
                                      required,
                                    },
                                  });
                                });
             }
            return validatonsArray;
          })(),
        ],

}

}

}

```




**HTML CODE**



```
<va-select
            v-for="(item, index) in addresses"
            :key="index"
            :error="
              v$.addresses[index].address_type.$invalid &&
              v$.addresses[index].address_type.$dirty
            "
            @blur="v$.addresses[index].address_type.$touch()"
            label="Address type"
            v-model="addresses[index].address_type"
            :required="true"
            :options="addressTypes"
          >
          </va-select>


```

**Comment by DwayneHawkins** [comment]
> Please update the docs 😣 I just spent half a day trying to figure out why $each wasn't working. I tried to find the page on github to send a pull request, but I don't think the vnext docs are in github?
> 
> I also agree that having to create extra components just because of vuelidate is burdensome, and that's exactly what I'm in the middle of right now.

Same here ... in 2022

**Comment by shentao** [comment]
Sorry to hear that. 
The docs already mention this in the Migration Guide: https://vuelidate-next.netlify.app/migration_guide.html
Here’s also the proposed alternative: https://vuelidate-next.netlify.app/advanced_usage.html#using-the-new-foreach-helper.

**Comment by lilz-sofia** [comment]
@shentao 
Hi!
Is the proposed alternative that you mentioned above also applicable for vue2? The syntax of the example seemed to be for vue3. 
https://vuelidate-next.netlify.app/advanced_usage.html#using-the-new-foreach-helper

I tried sth like this:
```
validations() {
  return {
    tasks: {
      $each: helpers.forEach({
      tracker: { required },
      description: { required },
      service_ids: { required }
      })
    }
  }
}
```
When I tried to access it with `v$.tasks.$each` but getting error. Is this the correct syntax?

**Comment by DevOrish** [comment]
> Sorry to hear that. The docs already mention this in the Migration Guide: https://vuelidate-next.netlify.app/migration_guide.html Here’s also the proposed alternative: https://vuelidate-next.netlify.app/advanced_usage.html#using-the-new-foreach-helper.


@shentao @dobromir-hristov 
Hi there,
I am currently migrating a huge app with a lot of forms, and the removal of $each has been a big headache for me.
I have read the entire thread and could feel the frustration about it. 
on the new forEach helper - there is no indication for $dirty, and we are using it heavily to allow good UX in our forms. 
am I bound to remove this property for good? is there something similar to it that im missing?

Anyway - I appreciate the care you give and the time to answer and implement. thank you
